src/backend/distributed/operations/citus_tools.c

```diff
@@ -52,14 +57,61 @@ static void ExecuteCommandsAndStoreResults(StringInfo *nodeNameArray,
 										   bool *statusArray,
 										   StringInfo *resultStringArray,
 										   int commandCount);
-static bool ExecuteRemoteQueryOrCommand(char *nodeName, uint32 nodePort,
-										char *queryString, StringInfo queryResult);
+static bool ExecuteRemoteQueryOrCommand(char *nodeName, uint32 nodePort, const char *user,
+										const char *database, char *queryString,
+										StringInfo queryResultString,
+										bool forceNewConnection);
 static Tuplestorestate * CreateTupleStore(TupleDesc tupleDescriptor,
 										  StringInfo *nodeNameArray, int *nodePortArray,
 										  bool *statusArray,
 										  StringInfo *resultArray, int commandCount);
 
 
+/*
+ * check_connection_to_node sends a simple query from a worker node to another
+ * node, and returns success status.
+ */
+Datum
+check_connection_to_node(PG_FUNCTION_ARGS)
+{
+	char *nodeName = PG_GETARG_TEXT_TO_CSTRING(0);
+	int32 nodePort = PG_GETARG_INT32(1);
+	char *userName = PG_GETARG_TEXT_TO_CSTRING_OR_NULL(2);
+	char *databaseName = PG_GETARG_TEXT_TO_CSTRING_OR_NULL(3);
+
+	bool success = CheckConnectionToNode(nodeName, nodePort, userName, databaseName);
+	PG_RETURN_BOOL(success);
+}
+
+
+/*
+ * CheckConnectionToNode sends a simple query from a node to another and returns success status
+ */
+static bool
+CheckConnectionToNode(char *nodeName, int32 nodePort, const char *userName,
+					  const char *databaseName)
+{
+	StringInfo queryResultString = makeStringInfo();
+	bool forceNewConnection = false;
+	bool success = false;
+
+	PG_TRY();
+	{
+		success =
+			ExecuteRemoteQueryOrCommand(nodeName, nodePort, userName, databaseName,
```

Comment:
Hmm, this API somehow terminates connection after the Tx, see that `connectionId` is changing (and you can attach debugger and break at `ShutdownConnection` function to verify):
```SQL
select check_connection_to_node('localhost', 9701) ;
NOTICE:  issuing SELECT 1
DETAIL:  on server onderkalaci@localhost:9701 connectionId: 10785
 check_connection_to_node 
--------------------------
 t
(1 row)

Time: 16.098 ms
[local] onderkalaci@postgres:5432-11076=# select check_connection_to_node('localhost', 9701) ;
NOTICE:  issuing SELECT 1
DETAIL:  on server onderkalaci@localhost:9701 connectionId: 10786
 check_connection_to_node 
--------------------------
```

The expectation would be this UDF do not terminate the connection.

It seems like ` AfterXactHostConnectionHandling()->ShouldShutdownConnection->RemoteTransactionIdle(connection) == 0`

triggers the connection to be terminated.  

I guess we should also call `ForgetResults` followed by `PQclear` in `ExecuteRemoteQueryOrCommand`. That'd ensure we consume all the pending activity, making sure we do not close the connection.

Can you think of adding a regression test for that?

Reply:
@hanefi did you check if the new API you used is effected by this one?

Reply:
I agree that I should have called `ForgetResults` here. However, as I switched to the new API.

`ExecuteOptionalRemoteCommand` always calls `ForgetResults` after a `PQclear` so we do not need to worry about that now.

My test results seem ok.

```
hanefi=# SELECT citus_check_connection_to_node('localhost', 9701);
LOG:  statement: SELECT citus_check_connection_to_node('localhost', 9701);
NOTICE:  issuing SELECT 1
DETAIL:  on server hanefi@localhost:9701 connectionId: 1
 citus_check_connection_to_node
--------------------------------
 t
(1 row)

hanefi=# SELECT citus_check_connection_to_node('localhost', 9702);
LOG:  statement: SELECT citus_check_connection_to_node('localhost', 9702);
NOTICE:  issuing SELECT 1
DETAIL:  on server hanefi@localhost:9702 connectionId: 2
 citus_check_connection_to_node
--------------------------------
 t
(1 row)

hanefi=# SELECT citus_check_connection_to_node('localhost', 9701);
LOG:  statement: SELECT citus_check_connection_to_node('localhost', 9701);
NOTICE:  issuing SELECT 1
DETAIL:  on server hanefi@localhost:9701 connectionId: 1
 citus_check_connection_to_node
--------------------------------
 t
(1 row)
```